File: FriTD_old/src/Entities/Enemy.java
package Entities;

import java.awt.Color;
import java.awt.Graphics2D;
import java.util.LinkedList;

import Enums.EnemyState;
import Helpers.MathHelper;
import UI.IDisplayableObject;

public class Enemy implements IDisplayableObject {
	
	private int id;
	private int x,y;
	private int hp;
	private int maxHp;
	private int speed;
	private int goldValue;
	private int HPCost;
	private EnemyState state;
	private PathSquare squareWayPoint;
	private LinkedList<Effect> effects;
	private int maxSpeed;
	
	
	

	public int getX() {
		return x;
	}


	
	public int getY() {
		return y;
	}



	public Enemy(int id, int x, int y, int hp, int speed, int goldValue, int hPCost, PathSquare squareWayPoint) {
		super();
		this.id = id;
		this.x = x;
		this.y = y;
		this.hp = hp;
		maxHp = hp;
		this.speed = speed;
		this.goldValue = goldValue;
		HPCost = hPCost;
		this.squareWayPoint = squareWayPoint;
		this.state = EnemyState.alive;
		this.effects = new LinkedList<>();
		this.maxSpeed = speed;
	}
	

	@Override
	public void display(Graphics2D g2d) {
		Color c = g2d.getColor();
		g2d.setColor(Color.black);
		g2d.fillOval(x-7, y-7, 15, 15);
		g2d.setColor(Color.red);
		g2d.fillRect(x-6, y-1, (int) (14*(hp*1.0/maxHp*1.0)), 2);
		g2d.setColor(c);
		
	}

	public void move() {
		LinkedList<Effect> expiredEffects = new LinkedList<>();;
		speed = maxSpeed;
		//add here projectile effect application
		for (Effect effect : effects) {
			speed = (int) ((int) speed*effect.getSlow());
			hp-=effect.getDmg();
			effect.reduceTTL();
			if(effect.getRemainingTurns()<=0)expiredEffects.addLast(effect);
		}
		if(hp<=0){
			this.state = EnemyState.dead;
			
		}
		for (Effect effect : expiredEffects) {
			effects.remove(effect);
		}
		
		if(state == EnemyState.dead)return;
		
		
		if(squareWayPoint == null){
			this.state = EnemyState.victorious;
			return;
		}
		
		
		int nextX, nextY;
		
		if(squareWayPoint.getX()!=x){
			if(squareWayPoint.getX() < x){
				nextX = x - speed;
			}else{
				nextX = x + speed;
			}
		}else{
			nextX = x;
		}
		if(squareWayPoint.getY()!=y){
			if(squareWayPoint.getY() < y){
				nextY = y - speed;
			}else{
				nextY = y + speed;
			}
		}else{
			nextY = y;
		}
		
		//MCH
		if(MathHelper.getDistanceBetweenPoints(nextX, nextY, squareWayPoint.getX(), squareWayPoint.getY())<2*speed){
			squareWayPoint = squareWayPoint.getNext();
			
		};
		
		x = nextX;
		y = nextY;
		
	}

	public boolean isDead() {
		
		return state == EnemyState.dead;
	}

	public int getGoldValue() {
		return goldValue;
	}

	public int getHPCost() {
		
		return HPCost;
	}

	public boolean isVictorious() {
		return state == EnemyState.victorious;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + id;
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Enemy other = (Enemy) obj;
		if (id != other.id)
			return false;
		return true;
	}



	public void applyEffect(Effect effect) {
		this.effects.addLast(effect);
	}

	
	
}

File: FriTD/TD/Entities/Enemy.cs
using System.Collections.Generic;
using System.Diagnostics;
using TD.Core;
using TD.Enums;
using TD.Helpers;

namespace TD.Entities
{
    public class Enemy:IDisplayableObject
    {
        public string Name { get; set; }
        public EnemyCategory Category { get; set; }
        public double Hp { get; set; }
        public double MaxHp { get; set; }
        public int Speed { get; set; }
        public int MaxSpeed { get; set; }
        public int Gold { get; set; }
        public int HpCost { get; set; }
        public int Armor { get; set; }
        public int MagicResist { get; set; }
        public int SpawnCd { get; set; }
        public EnemyState State { get; set; }
        private readonly LinkedList<Effect> _effects;
        public PathSquare SquareWayPoint { get; set; }

        public int X { get; set; }
        public int Y { get; set; }
        public int Id { get; set; }
        public double Perc
        {
            get {return Hp/MaxHp;}
            set { Perc = value; }
        }

        public Enemy()
        {
            _effects = new LinkedList<Effect>();
        }


        public void Move()
        {
            Debug.WriteLine("Enemy hp: "+Hp+" type "+Id+ " moved from "+Y+" "+X+" "+Speed);
            
            var expiredEffects = new LinkedList<Effect>();
            Speed = MaxSpeed;

            //add here projectile effect application
            foreach (var effect in _effects)
            {
                Speed = (int)(Speed * ((100-effect.Slow)/100));
                Hp -= effect.Dmg(Armor, MagicResist);
                effect.ReduceTtl();
                if (effect.RemainingTurns <= 0) expiredEffects.AddLast(effect);
            }
            if (Hp <= 0)
            {
                State = EnemyState.Dead;

            }
            foreach (var effect in expiredEffects)
            {
                _effects.Remove(effect);
            }

            if (State == EnemyState.Dead) return;


            if (SquareWayPoint == null)
            {
                State = EnemyState.Victorious;
                return;
            }


            int nextX, nextY;

            if (SquareWayPoint.X != X)
            {
                if (SquareWayPoint.X < X)
                {
                    nextX = X - Speed;
                }
                else
                {
                    nextX = X + Speed;
                }
            }
            else
            {
                nextX = X;
            }
            if (SquareWayPoint.Y != Y)
            {
                if (SquareWayPoint.Y < Y)
                {
                    nextY = Y - Speed;
                }
                else
                {
                    nextY = Y + Speed;
                }
            }
            else
            {
                nextY = Y;
            }
            Debug.WriteLine("Enemy hp: " + Hp + " type " + Id + " moved to " + Y + " " + X);
            Debug.WriteLine("Way point "+SquareWayPoint.X +" "+SquareWayPoint.Y);
            //MCH
            if (MathHelper.DistanceBetweenPoints(nextX, nextY, SquareWayPoint.X, SquareWayPoint.Y) < 2 * Speed)
            {
                SquareWayPoint = SquareWayPoint.Next;

            }

            X = nextX;
            Y = nextY;

        }

        public bool IsDead()
        {

            return State == EnemyState.Dead;
        }
        

        public bool IsVictorious()
        {
            return State == EnemyState.Victorious;
        }

        




        public void ApplyEffect(Effect effect)
        {
            _effects.AddLast(effect);
        }

        
    }
}